cpp/include/rapidsmpf/streaming/core/memory_reserve_or_wait.hpp

```diff
@@ -0,0 +1,175 @@
+  public:
+    /**
+     * @brief Constructs a `MemoryReserveOrWait` instance.
+     *
+     * @param mem_type The memory type for which reservations are requested.
+     * @param ctx Streaming context.
+     * @param timeout Optional timeout duration. This timeout applies to how long pending
+     * requests may wait without making progress. If the timeout expires, a
+     * `reserve_or_wait()` returns even if no memory became available. If not explicitly
+     * provided, the timeout is read from the option key `"memory_reserve_timeout_ms"`,
+     * which defaults to 100 ms.
+     */
+    MemoryReserveOrWait(
+        MemoryType mem_type,
+        std::shared_ptr<Context> ctx,
+        std::optional<Duration> timeout = std::nullopt
+    );
+
+    ~MemoryReserveOrWait() noexcept;
+
+    /**
+     * @brief Shuts down all pending memory reservation requests.
+     *
+     * @return A coroutine that completes only after all pending requests have been
+     * cancelled and the periodic memory check task has exited.
+     */
+    Node shutdown();
+
+    /**
+     * @brief Attempts to reserve memory or waits until the reservation can be satisfied.
+     *
+     * This coroutine submits a memory reservation request and then suspends until
+     * either sufficient memory becomes available or no progress is made within the
+     * configured timeout.
+     *
+     * If the timeout expires before the request can be fulfilled, an empty
+     * `MemoryReservation` is returned.
+     *
+     * @param size Number of bytes to reserve.
+     * @param future_release_potential Estimated number of bytes the requester may release
+     * in the future, used as a heuristic when selecting which eligible request to satisfy
+     * first.
+     * @return A `MemoryReservation` representing the allocated memory, or an empty
+     * reservation if the timeout expires.
+     *
+     * @throws std::runtime_error If shutdown occurs before the request can be processed.
+     */
+    coro::task<MemoryReservation> reserve_or_wait(
+        std::size_t size, std::size_t future_release_potential
+    );
+
+    /**
+     * @brief Returns the number of pending memory reservation requests.
+     *
+     * It may change concurrently as requests are added or fulfilled.
+     *
+     * @return The number of outstanding reservation requests.
+     */
+    [[nodiscard]] std::size_t size() const;
+
+    /**
+     * @brief Returns the number of iterations performed by `periodic_memory_check()`.
+     *
+     * This counter is incremented once per loop iteration inside
+     * `periodic_memory_check()`, and can be useful for diagnostics or testing.
+     *
+     * @return The total number of memory-check iterations executed so far.
+     */
+    [[nodiscard]] std::size_t periodic_memory_check_counter() const;
+
+  private:
+    /**
+     * @brief Represents a single memory reservation request.
+     *
+     * A `ResReq` is inserted into a sorted container and processed by
```

Comment:
*nit*: Can we call this a `ReservationRequest` (or perhaps `Request` since it's a nested class)?